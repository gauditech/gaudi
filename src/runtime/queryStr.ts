import { source } from "common-tags";
import _ from "lodash";

import { getRef, getTargetModel } from "@src/common/refs";
import { ensureEqual } from "@src/common/utils";
import { BinaryOperator } from "@src/types/ast";
import {
  Definition,
  FilterDef,
  ModelDef,
  QueryDef,
  QueryDefPath,
  SelectableItem,
} from "@src/types/definition";

export function queryToString(def: Definition, q: QueryDef): string {
  switch (q.from.kind) {
    case "model": {
      const { value: model } = getRef<"model">(def, q.from.refKey);
      return source`
      SELECT ${selectToString(def, q.select)}
      FROM "${model.dbname}" AS ${toAlias([model.name])}
      ${q.joinPaths.map((j) => joinToString(def, j))}
      WHERE ${filterToString(q.filter)}`;
    }
    case "query":
      return source`
      SELECT ${selectToString(def, q.select)}
      FROM (${queryToString(def, q.from.query)}) AS ${toAlias(q.from.query.fromPath)}
      ${q.joinPaths.map((j) => joinToString(def, j))}
      WHERE ${filterToString(q.filter)}`;
  }
}

function selectToString(def: Definition, select: SelectableItem[]) {
  return select
    .map((item) => {
      switch (item.kind) {
        case "field": {
          const { value: field } = getRef<"field">(def, item.refKey);
          return `${toAlias(_.initial(item.namePath))}.${field.dbname} AS ${item.alias}`;
        }
        case "constant": {
          // FIXME security issue, this is not an escaped value!!
          ensureEqual(item.type, "integer");
          return `${item.value} AS "${item.alias}"`;
        }
      }
    })
    .join(", ");
}

function toAlias(np: string[]): string {
  return `"${np.join(".")}"`;
}

function filterToString(filter: FilterDef): string {
  if (filter === undefined) return "TRUE = TRUE";
  switch (filter.kind) {
    case "literal": {
      switch (filter.type) {
        case "boolean":
          return filter.value ? "TRUE" : "FALSE";
        case "null":
          return "NULL";
        case "text":
          return `'${filter.value}'`;
        case "integer":
          return filter.value.toString();
      }
    }
    // Due to bug in eslint/prettier, linter complains that `break` is expected in the case "literal"
    // Since inner switch is exaustive, break is unreachable so prettier deletes it
    // eslint-disable-next-line no-fallthrough
    case "alias": {
      const np = filter.namePath.slice(0, filter.namePath.length - 1);
      const f = filter.namePath.at(filter.namePath.length - 1);
      return `${toAlias(np)}.${f}`;
    }
    case "binary": {
      const fstr = `${filterToString(filter.lhs)} ${opToString(filter.operator)} ${filterToString(
        filter.rhs
      )}`;
      if (filter.operator === "or") {
        return `(${fstr})`;
      }
      return fstr;
    }
    case "variable": {
      return `:${filter.name}`;
    }
  }
}

function opToString(op: BinaryOperator): string {
  switch (op) {
    case "is":
      return "=";
    case "is not":
      return "<>";
    default:
      return op.toUpperCase();
  }
}

function getQuerySource(def: Definition, q: QueryDef): ModelDef {
  switch (q.from.kind) {
    case "model":
      return getRef<"model">(def, q.from.refKey).value;
    case "query": {
      return getQuerySource(def, q.from.query);
    }
  }
}

function joinToString(def: Definition, join: QueryDefPath): string {
  const model = getTargetModel(def.models, join.refKey);
  const joinNames = getJoinNames(def, join.refKey);
  const joinMode = join.joinType === "inner" ? "JOIN" : "LEFT JOIN";
  const joinFilter: FilterDef = {
    kind: "binary",
    operator: "is",
    lhs: { kind: "alias", namePath: [..._.initial(join.namePath), joinNames.from] },
    rhs: { kind: "alias", namePath: [...join.namePath, joinNames.to] },
  };

  let src: string;
  if (join.kind === "query") {
    const { value: query } = getRef<"query">(def, join.refKey);
    const sourceModel = getQuerySource(def, query);
    // extend select
    const conn: SelectableItem = {
      kind: "field",
      refKey: `${sourceModel.name}.id`,
      alias: '"__join_connection"',
      name: "id",
      namePath: [sourceModel.name, "id"],
    };
    const retModel = getRef<"model">(def, query.retType).value;
    const fields = retModel.fields.map(
      (f): SelectableItem => ({
        kind: "field",
        refKey: f.refKey,
        alias: f.name,
        name: f.name,
        namePath: [...query.fromPath, f.name],
      })
    );
    src = source`(
      ${queryToString(def, { ...query, select: [...query.select, ...fields, conn] })})`; // FIXME should remove query.select?
  } else {
    src = `"${model.dbname}"`;
  }

  return source`
  ${joinMode}
  ${src} AS ${toAlias(join.namePath)}
  ON ${filterToString(joinFilter)}
  ${join.joinPaths.map((j) => joinToString(def, j))}`;
}

function getJoinNames(def: Definition, refKey: string): { from: string; to: string } {
  const prop = getRef(def, refKey);
  switch (prop.kind) {
    case "reference": {
      const reference = prop.value;
      const { value: field } = getRef<"field">(def, reference.fieldRefKey);
      const { value: refField } = getRef<"field">(def, reference.toModelFieldRefKey);

      return { from: field.name, to: refField.name };
    }
    case "relation": {
      const relation = prop.value;
      const { value: reference } = getRef<"reference">(def, relation.throughRefKey);
      const { value: field } = getRef<"field">(def, reference.fieldRefKey);
      const { value: refField } = getRef<"field">(def, reference.toModelFieldRefKey);

      return { to: field.name, from: refField.name };
    }
    case "query": {
      return { to: '"__join_connection"', from: "id" };
    }
    default:
      throw new Error(`Kind ${prop.kind} not supported`);
  }
}
