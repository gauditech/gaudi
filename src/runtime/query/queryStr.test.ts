import _ from "lodash";

import { endpointQueries } from "./buildQuery";
import { queryToString } from "./queryStr";

import { compile, compose, parse } from "@src/index";

describe("Endpoint queries", () => {
  it("nested query", () => {
    const bp = `
    model Org {
      field slug { type text, unique }
      relation repos { from Repo, through org }
      query public_repos { from repos, filter { is_public is true }}
    }
    model Repo {
      reference org { to Org }
      field is_public { type boolean }
      field name { type text }
    }
    entrypoint Orgs {
      target model Org
      identify with slug
      entrypoint Repos {
        target relation public_repos
        response { id, name }

        get endpoint {}
      }
    }

    `;
    const def = compose(compile(parse(bp)));

    const endpoint = def.entrypoints[0].entrypoints[0].endpoints[0];
    const q = endpointQueries(def, endpoint);
    expect(q).toMatchSnapshot();
  });
  it("chained nested query", () => {
    const bp = `
    model Org {
      field slug { type text, unique }
      relation repos { from Repo, through org }
      query public_repos { from repos, filter { is_public is true }}
      query public_issues { from public_repos.issues  }
    }
    model Repo {
      reference org { to Org }
      field is_public { type boolean }
      field name { type text }
      relation issues { from Issue, through repo }
    }
    model Issue {
      reference repo { to Repo }
      field name { type text }
    }
    entrypoint Orgs {
      target model Org
      identify with slug
      entrypoint RepoIssues {
        target relation public_issues
        response { id, name }
        get endpoint {}
      }
    }

    `;
    const def = compose(compile(parse(bp)));
    const endpoint = def.entrypoints[0].entrypoints[0].endpoints[0];
    const q = endpointQueries(def, endpoint);
    expect(q).toMatchSnapshot();
  });
});
