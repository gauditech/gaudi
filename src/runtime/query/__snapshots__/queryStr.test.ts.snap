// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Endpoint queries chained nested query 1`] = `
{
  "target": {
    "alias": "target",
    "name": "target",
    "query": {
      "filter": {
        "kind": "binary",
        "lhs": {
          "kind": "binary",
          "lhs": {
            "kind": "alias",
            "namePath": [
              "Org",
              "slug",
            ],
          },
          "operator": "is",
          "rhs": {
            "kind": "variable",
            "name": "org_slug",
            "type": "text",
          },
        },
        "operator": "and",
        "rhs": {
          "kind": "binary",
          "lhs": {
            "kind": "alias",
            "namePath": [
              "Org",
              "public_issues",
              "id",
            ],
          },
          "operator": "is",
          "rhs": {
            "kind": "variable",
            "name": "issue_id",
            "type": "integer",
          },
        },
      },
      "from": {
        "kind": "model",
        "refKey": "Org",
      },
      "fromPath": [
        "Org",
        "public_issues",
      ],
      "joinPaths": [
        {
          "joinPaths": [],
          "joinType": "inner",
          "kind": "query",
          "name": "public_issues",
          "namePath": [
            "Org",
            "public_issues",
          ],
          "refKey": "Org.public_issues",
          "retType": "Issue",
        },
      ],
      "name": "target",
      "nullable": false,
      "refKey": "N/A",
      "retType": "Issue",
      "select": [
        {
          "alias": "id",
          "kind": "field",
          "name": "id",
          "namePath": [
            "Org",
            "public_issues",
            "id",
          ],
          "refKey": "Issue.id",
        },
        {
          "alias": "name",
          "kind": "field",
          "name": "name",
          "namePath": [
            "Org",
            "public_issues",
            "name",
          ],
          "refKey": "Issue.name",
        },
      ],
    },
    "related": [],
  },
}
`;

exports[`Endpoint queries nested query 1`] = `
{
  "target": {
    "alias": "target",
    "name": "target",
    "query": {
      "filter": {
        "kind": "binary",
        "lhs": {
          "kind": "binary",
          "lhs": {
            "kind": "alias",
            "namePath": [
              "Org",
              "slug",
            ],
          },
          "operator": "is",
          "rhs": {
            "kind": "variable",
            "name": "org_slug",
            "type": "text",
          },
        },
        "operator": "and",
        "rhs": {
          "kind": "binary",
          "lhs": {
            "kind": "alias",
            "namePath": [
              "Org",
              "public_repos",
              "id",
            ],
          },
          "operator": "is",
          "rhs": {
            "kind": "variable",
            "name": "repo_id",
            "type": "integer",
          },
        },
      },
      "from": {
        "kind": "model",
        "refKey": "Org",
      },
      "fromPath": [
        "Org",
        "public_repos",
      ],
      "joinPaths": [
        {
          "joinPaths": [],
          "joinType": "inner",
          "kind": "query",
          "name": "public_repos",
          "namePath": [
            "Org",
            "public_repos",
          ],
          "refKey": "Org.public_repos",
          "retType": "Repo",
        },
      ],
      "name": "target",
      "nullable": false,
      "refKey": "N/A",
      "retType": "Repo",
      "select": [
        {
          "alias": "id",
          "kind": "field",
          "name": "id",
          "namePath": [
            "Org",
            "public_repos",
            "id",
          ],
          "refKey": "Repo.id",
        },
        {
          "alias": "name",
          "kind": "field",
          "name": "name",
          "namePath": [
            "Org",
            "public_repos",
            "name",
          ],
          "refKey": "Repo.name",
        },
      ],
    },
    "related": [],
  },
}
`;
